lib/livebook/notebook/explore/intro_to_nx.livemd

```diff
@@ -1,3 +1,742 @@
+
+  defn multiply(x, y) do
+    inspect_expr(x * y)
+  end
+end
+```
+
+This code creates a module with a `defn` block, and will compile to
+the `XLA` backend. You'll see the expression when you evaluate it:
+
+```elixir
+VerboseMath.multiply(Nx.tensor([1, 2]), 3)
+```
+
+Now, you can see the intermediate form. Notice you didn't see the
+compiled form until the code was evaluated the first time. The first
+part of the result is the expression. It has two parameters,  `a` and
+`b`. These two bindings are the inbound arguments of the function,
+but instead of referring to them as `x` and `y`, the expression
+arbitrarily labels these with `a` and `b`. They are bound to
+a `s64[2]` tensor and a `s64` scalar. The result is a third binding
+called `c`.
+
+This intermediate AST is what Nx hands off to a compiler. It also
+forms the foundations of autograd.
+
+## Automatic differentiation (autograd)
+
+An important mathematical property for a function is the
+rate of change, or the gradient. These gradients are critical
+for solving systems of equations and building probablistic
+models. In advanced math, derivatives, or differential equations,
+are used to take gradients. Nx can compute these derivatives
+automatically through a feature called automatic differentiation,
+or autograd.
+
+Here's how it works.
+
+```elixir
+h(Nx.Defn.grad())
+```
+
+We'll build a module with a few functions,
+and then create another function to create the gradients of those
+functions. The function `grad/1` takes a function, and returns
+a function returning the gradient. We have two functions: `poly/1`
+is a simple numerical definitin, and `poly_slope_at/1` returns
+its gradient:
+
+<!-- livebook:{"break_markdown":true} -->
+
+$$
+  poly: f(x) = 3x^2 + 2x + 1 \\
+$$
+
+<!-- livebook:{"break_markdown":true} -->
+
+$$
+  polySlopeAt: g(x) = 6x + 2
+$$
+
+<!-- livebook:{"break_markdown":true} -->
+
+Here's the Elixir equivalent of those functions:
+
+```elixir
+defmodule Funs do
+  defn poly(x) do
+    3 * Nx.power(x, 2) + 2 * x + 1
+  end
+
+  def poly_slope_at(x) do
+    grad(&poly/1).(x)
+  end
+end
+```
+
+Notice the second `defn`. It uses `grad/1` to take its
+derivative using autograd. It uses the intermediate `defn` AST
+and mathematical composition to compute the derivative. You can
+see it at work here:
+
+```elixir
+Funs.poly_slope_at(2)
+```
+
+Nice. If you plug the number 2 into the function $$ 6x + 2 $$
+you get 14! Said another way, if you look at the graph at
+exactly 2, the rate of increase is 14 units of `poly(x)`
+for every unit of `x`, precisely at `x`.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Nx also has helpers to get gradients corresponding to a number of inputs.
+These come into play when solving systems of equations.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Now, you try. Find a function computing the gradient of a `sin` wave.
+
+```elixir
+# your code here
+```
+
+In the notebooks to follow, we'll use these techniques to
```

Comment:
@josevalim do we want multiple Nx notebooks in the explore section and links to such, or rather have a single introduction notebook and link to more notebooks in the Nx repo (once there are such)?

Reply:
All sounds good. I will wait for @josevalim to weigh in and make all the changes at once. My intuition is that it would be nice to have two more notebooks in this area: one on linear regression, and one on tensor creation/operators/broadcasting/slicing since those are the basic building blocks folks will need to work in nx. I am happy to create those if this is a good direction for the project.

Reply:
If it's desirable, I think a second notebook with tensors, defns, operators, slicing, and more exercises would be pretty useful. And Sean and I definitely talked about doing a linear regression, and I think that would be a cool notebook that's a bit more interactive.